src/Core/Rendering/RenderController.cpp

```diff
@@ -754,13 +763,75 @@ namespace MxEngine
 
         std::swap(inputOutput, temporary0);
     }
-    void RenderController::ApplyHDRToLDRConversion(CameraUnit& camera, TextureHandle& input, TextureHandle& output)
+
+    void RenderController::ApplyLensFlare(CameraUnit& camera, TextureHandle& input, TextureHandle& temporaryQuater0, TextureHandle& temporaryQuater1, TextureHandle& temporaryQuater2, TextureHandle& temporary1)
     {
+        if (camera.LensFlare == nullptr)
+            return;
+
+        MAKE_SCOPE_PROFILER("RenderController::ApplyLensFlare()");
+
+        //TODO(fall2019): support chromatic aberration  
+        //TODO(fall2019): add starbust effect 
+        float scale = camera.LensFlare->GetLensFlareScale();
+        float bias = camera.LensFlare->GetLensFlareBias();
+        int numOfGhosts = camera.LensFlare->GetLensFlareNumOfGhosts();
+        float dispersal = camera.LensFlare->GetLensFlareGhostDispersal();
+        float haloWidth = camera.LensFlare->GetLensFlareHaloWidth();
+
+        input->GenerateMipmaps();
+
+        //calc ghosts
+        {
+            temporaryQuater0->GenerateMipmaps();
+            auto& shaderGhost = this->Pipeline.Environment.Shaders["LensFlareGhosts"_id];
+            shaderGhost->Bind();
+            shaderGhost->SetUniform("uScale", scale);
+            shaderGhost->SetUniform("uBias", bias);
+            input->Bind(0);
+            camera.AverageWhiteTexture->Bind(1);
+            this->RenderToTextureNoClear(temporaryQuater0, shaderGhost);
+            this->ApplyGaussianBlur(temporaryQuater0, temporaryQuater1, 3);
+            temporaryQuater1->GenerateMipmaps();
+        }
+
+        //calc halo
+        {
+            auto& shaderHalo = this->Pipeline.Environment.Shaders["LensFlareHalo"_id];
+            shaderHalo->Bind();
+            shaderHalo->SetUniform("uGhostDispersal", dispersal);
+            shaderHalo->SetUniform("uHaloWidth", haloWidth);
+            int mipLevel = ceil(Log2(Max(temporaryQuater1->GetWidth(), temporaryQuater1->GetHeight()))) - 2;
```

Comment:
log(8) = 3, mips for 8: 8, 4, 2, 1, total: 4
log(10) = 3.3 (round to 4), mips for 10: 10, 5, 2, 1, total: 4
so I guess total mips is floor(log(X)) + 1, not ceil

https://stackoverflow.com/questions/9572414/how-many-mipmaps-does-a-texture-have-in-opengl